src/plugins/colum-chooser/column-chooser.component.ts

```diff
@@ -0,0 +1,231 @@
+import {Component, Optional, OnInit, OnDestroy, Input, Output, EventEmitter, ChangeDetectorRef} from '@angular/core';
+import {Command} from 'ng2-qgrid/core/command';
+import {PipeUnit} from 'ng2-qgrid/core/pipe/pipe.unit2';
+import {GridService} from 'ng2-qgrid/main/grid';
+import * as columnService from 'ng2-qgrid/core/column/column.service';
+import {isFunction, noop} from 'ng2-qgrid/core/utility';
+import {Aggregation} from 'ng2-qgrid/core/services';
+
+import {RootService} from 'ng2-qgrid/infrastructure/component';
+import {PluginComponent} from '../plugin.component';
+import {TemplatePath} from 'ng2-qgrid/core/template';
+
+const GRID = 'qGrid';
+const COLUMN_CHOOSER_NAME = `${GRID}ColumnChooser`;
+
+@Component({
+	selector: 'q-grid-column-chooser',
+	template: `
+		<ng-container key="column-chooser.tpl" [context]="context"></ng-container>
+	`
+	// templateUrl: './column-chooser.tpl.html'
+})
+export class ColumnChooserComponent extends PluginComponent implements OnInit, OnDestroy {
+	@Input('canAggregate') columnChooserCanAggregate: boolean;
+	@Output('submit') submitEvent = new EventEmitter<any>();
+	@Output('cancel') cancelEvent = new EventEmitter<any>();
+
+	private service: any;
+	private aggregations: any;
+	private origin: any;
+	private filter: string = '';
+
+	private columns: any[];
+
+	constructor(@Optional() root: RootService, private gridService: GridService, private cdRef: ChangeDetectorRef) {
```

Comment:
cdRef why?

Reply:
We still have a bug here that hasn't been fixed yet: while a popup window is being opened with the Column Chooser on it the following kind of exception is thrown occasionally (but quite often):
 ThemeComponent.html:316 ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'false'. Current value: 'true'.
I used to use this to work around this sort of issue, so it might be needed in this case as well. I am still investigating the problem...

Reply:
fixed (cdRef removed). The bug mentioned above is still there